provider/acme/provider.go

```diff
@@ -0,0 +1,479 @@
+	HTTPChallenge *HTTPChallenge `description:"Activate HTTP-01 Challenge"`
+
+	Domains []types.Domain `description:"SANs (alternative domains) to each main domain using format: --acme.domains='main.com,san1.com,san2.com' --acme.domains='main.net,san1.net,san2.net'"`
+
+	Store        Store
+	certificates []*Certificate
+	account      *Account
+
+	client            *acme.Client
+	certsChan         chan *Certificate
+	configurationChan chan<- types.ConfigMessage
+	dynamicCerts      *safe.Safe
+	clientMutex       sync.Mutex
+
+	configFromListenerChan chan types.Configuration
+}
+
+// Certificate is a struct which contains all data needed from an ACME certificate
+type Certificate struct {
+	Domain      types.Domain
+	Certificate []byte
+	Key         []byte
+}
+
+// DNSChallenge contains DNS challenge Configuration
+type DNSChallenge struct {
+	Provider         string         `description:"Use a DNS-01 based challenge provider rather than HTTPS."`
+	DelayBeforeCheck flaeg.Duration `description:"Assume DNS propagates after a delay in seconds rather than finding and querying nameservers."`
+}
+
+// HTTPChallenge contains HTTP challenge Configuration
+type HTTPChallenge struct {
+	EntryPoint string `description:"HTTP challenge EntryPoint"`
+}
+
+// SetConfigListenerChan initializes the configFromListenerChan
+func (p *Provider) SetConfigListenerChan(configFromListenerChan chan types.Configuration) {
+	p.configFromListenerChan = configFromListenerChan
+}
+
+func (p *Provider) init() error {
+	if p.ACMELogging {
+		acme.Logger = fmtlog.New(os.Stderr, "legolog: ", fmtlog.LstdFlags)
+	} else {
+		acme.Logger = fmtlog.New(ioutil.Discard, "", 0)
+	}
+
+	var err error
+	if p.Store == nil {
+		err = errors.New("no store found for the ACME provider")
+		log.Error(err)
+		return err
+	}
+
+	p.account, err = p.Store.GetAccount()
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+
+	p.certificates, err = p.Store.GetCertificates()
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+
+	p.watchCertificate()
+	p.watchNewDomains()
+
+	return nil
+}
+
+func (p *Provider) initAccount() (*Account, error) {
+	if p.account == nil || len(p.account.Email) == 0 {
+		var err error
+		p.account, err = NewAccount(p.Email)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return p.account, nil
+}
+
+// ListenConfiguration sets a new configuration into the configFromListenerChan
+func (p *Provider) ListenConfiguration(config types.Configuration) {
+	p.configFromListenerChan <- config
+}
+
+// ListenRequest resolves new certificates for a domain from an incoming request and retrun a valid Certificate to serve (onDemand option)
+func (p *Provider) ListenRequest(domain string) (*tls.Certificate, error) {
+	acmeCert, err := p.resolveCertificate(types.Domain{Main: domain})
+	if err != nil {
+		return nil, err
+	}
+	certificate, err := tls.X509KeyPair(acmeCert.Certificate, acmeCert.PrivateKey)
+	return &certificate, err
+}
+
+func (p *Provider) watchNewDomains() {
+	safe.Go(func() {
```

Comment:
![poupees](https://user-images.githubusercontent.com/6207234/36388778-44903f04-159d-11e8-9c52-09f32a753f12.jpeg)

Reply:
Yes lot of `go routines`...  :stuck_out_tongue: